src/lib/x509/x509_ext.cpp

```diff
@@ -868,6 +877,831 @@ const std::string& TNAuthList::Entry::telephone_number() const {
+   // This process is repeated until there are no further changes, because multiple adjacent elements may need to be combinded one by one.
+
+   if(!ranges.has_value()) {
+      return std::nullopt;
+   }
+
+   std::vector<T> current(ranges.value().begin(), ranges.value().end());
+
+   if(current.empty()) {
+      return current;
+   }
+   bool changed = true;
+   while(changed) {
+      changed = false;
+      // sort by size of the min value
+      std::sort(current.begin(), current.end(), [](T& a, T& b) { return a.min() < b.min(); });
+      for(size_t i = 0; i < current.size() - 1; i++) {
+         const T a = current[i];
+         const T b = current[i + 1];
+         // they are either adjacent or overlap
+         if(b.min() <= a.max() || b.min() == (a.max() + 1)) {
+            // erase old a and b
+            current.erase(current.begin() + i, current.begin() + i + 2);
+            current.insert(current.begin() + i, T(a.min(), std::max(a.max(), b.max())));
+            // look again at the new element we just created, might need to be merged with others still
+            i -= 1;
+            changed = true;
+         }
+      }
+   }
+   return current;
+}
+
+template <typename T>
+bool validate_subject_in_issuer(const std::vector<T>& subject, const std::vector<T>& issuer) {
+   // ensures that the subject ranges are enclosed by the issuer ranges
+   // both vectors are already sorted, so we can do this in O(n+m)
+
+   // the issuer has 0 ranges to validate against, so this can only work if the subject also has none
+   if(issuer.size() == 0) {
+      return subject.size() == 0;
+   }
+   for(auto subj = subject.begin(), issu = issuer.begin(); subj != subject.end();) {
+      // the subject is outside of the closest issuer range on the left (min) side
+      if(subj->min() < issu->min()) {
+         return false;
+      }
+      // the issuer range is smaller than the subject range, step to the next issuer range
+      if(subj->min() > issu->max()) {
+         issu++;
+         // we have run out of issuer ranges, but still have subject ranges left to validate
+         if(issu == issuer.end() && subj != subject.end()) {
+            return false;
+         }
+         continue;
+      }
+      // the subject is outside of the closest issuer range on the right (max) side
+      if(subj->max() > issu->max()) {
+         return false;
+      }
+      subj++;
+   }
+   return true;
+}
+
+template <IPAddressBlocks::Version V>
+void populate_validation_map(uint32_t afam,
+                             const IPAddressBlocks::IPAddressFamily::AddrChoice& choice,
+                             IPValidationMap<V>& map) {
+   const std::optional<IPRangeVec<V>>& ranges = std::get<IPAddressBlocks::IPAddressChoice<V>>(choice).ranges();
+   bool has_value = ranges.has_value();
+   const IPRangeVec<V>* value = has_value ? &ranges.value() : nullptr;
+   map.emplace(afam, std::make_pair(has_value, std::move(value)));
```

Comment:
It's not immediately clear what a `nullptr` means in an `IPValidationMap`. A comment near the type definition above would be helpful.